File: it-agents/skills/code-architect/references/code-examples/python/abstract_factory.py
"""Abstract Factory — create families of related objects without specifying concrete classes."""

from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def render(self) -> str:
        ...


class Checkbox(ABC):
    @abstractmethod
    def render(self) -> str:
        ...


class LightButton(Button):
    def render(self) -> str:
        return "[ Light Button ]"


class LightCheckbox(Checkbox):
    def render(self) -> str:
        return "[x] Light Checkbox"


class DarkButton(Button):
    def render(self) -> str:
        return "[ Dark Button ]"


class DarkCheckbox(Checkbox):
    def render(self) -> str:
        return "[x] Dark Checkbox"


class WidgetFactory(ABC):
    """Each concrete factory produces a coherent family of widgets."""

    @abstractmethod
    def create_button(self) -> Button:
        ...

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        ...


class LightFactory(WidgetFactory):
    def create_button(self) -> Button:
        return LightButton()

    def create_checkbox(self) -> Checkbox:
        return LightCheckbox()


class DarkFactory(WidgetFactory):
    def create_button(self) -> Button:
        return DarkButton()

    def create_checkbox(self) -> Checkbox:
        return DarkCheckbox()


def build_ui(factory: WidgetFactory) -> str:
    return f"{factory.create_button().render()} {factory.create_checkbox().render()}"


if __name__ == "__main__":
    print(build_ui(LightFactory()))
    print(build_ui(DarkFactory()))

File: it-agents/skills/code-architect/references/code-examples/php/AbstractFactory.php
<?php
// Abstract Factory — create families of related objects without specifying concrete classes.
declare(strict_types=1);

interface Button
{
    public function render(): string;
}

interface Checkbox
{
    public function toggle(): string;
}

interface WidgetFactory
{
    public function makeButton(): Button;
    public function makeCheckbox(): Checkbox;
}

final class LightButton implements Button
{
    public function render(): string { return '[ Light Button ]'; }
}

final class LightCheckbox implements Checkbox
{
    public function toggle(): string { return '( ) light checkbox'; }
}

final class DarkButton implements Button
{
    public function render(): string { return '[ Dark Button ]'; }
}

final class DarkCheckbox implements Checkbox
{
    public function toggle(): string { return '(x) dark checkbox'; }
}

final class LightThemeFactory implements WidgetFactory
{
    public function makeButton(): Button { return new LightButton(); }
    public function makeCheckbox(): Checkbox { return new LightCheckbox(); }
}

final class DarkThemeFactory implements WidgetFactory
{
    public function makeButton(): Button { return new DarkButton(); }
    public function makeCheckbox(): Checkbox { return new DarkCheckbox(); }
}

function buildScreen(WidgetFactory $factory): void
{
    echo $factory->makeButton()->render() . ' ' . $factory->makeCheckbox()->toggle() . PHP_EOL;
}

// Demo
buildScreen(new LightThemeFactory());
buildScreen(new DarkThemeFactory());
